core/arch/riscv64/mangle.c

```diff
@@ -385,13 +394,406 @@ mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+    scratch_reg2 = pick_scratch_reg(dcontext, instr, scratch_reg1, &slot2);
+
+    /* Spill scratch registers. */
+    PRE(ilist, instr, instr_create_save_to_tls(dcontext, scratch_reg1, slot1));
+    PRE(ilist, instr, instr_create_save_to_tls(dcontext, scratch_reg2, slot2));
+
+    /* Restore address saved by exclusive load and check if it's equal to src0. */
+    PRE(ilist, instr,
+        instr_create_restore_from_tls(dcontext, scratch_reg1, TLS_LRSC_ADDR_SLOT));
+    PRE(ilist, instr,
+        INSTR_CREATE_bne(dcontext, opnd_create_instr(fail), opnd_create_reg(scratch_reg1),
+                         src0));
+
+    /* Restore size saved by exclusive load and check if it's equal to current size. */
+    PRE(ilist, instr,
+        instr_create_restore_from_tls(dcontext, scratch_reg1, TLS_LRSC_SIZE_SLOT));
+    PRE(ilist, instr,
+        XINST_CREATE_load_int(dcontext, opnd_create_reg(scratch_reg2),
+                              opnd_add_flags(opnd_create_immed_int(opsz, OPSZ_12b),
+                                             DR_OPND_IMM_PRINT_DECIMAL)));
+    PRE(ilist, instr,
+        INSTR_CREATE_bne(dcontext, opnd_create_instr(fail), opnd_create_reg(scratch_reg1),
+                         opnd_create_reg(scratch_reg2)));
+
+    PRE(ilist, instr,
+        instr_create_restore_from_tls(dcontext, scratch_reg1, TLS_LRSC_VALUE_SLOT));
+    PRE(ilist, instr, loop);
+
+    /* Convert exclusive store to a compare-and-swap.
+     * Begin of the LR/SC sequence.
+     */
+    opcode = instr_get_opcode(instr) == OP_sc_d ? OP_lr_d : OP_lr_w;
+    PRE(ilist, instr,
+        instr_create_1dst_2src(dcontext, opcode, opnd_create_reg(scratch_reg2), src0,
+                               opnd_create_immed_int(0b11, OPSZ_2b)));
+    PRE(ilist, instr,
+        INSTR_CREATE_bne(dcontext, opnd_create_instr(end), opnd_create_reg(scratch_reg1),
+                         opnd_create_reg(scratch_reg2)));
+
+    /* instr is here. */
+
+    PRE(ilist, next_instr,
+        INSTR_CREATE_bne(dcontext, opnd_create_instr(loop), dst,
+                         opnd_create_reg(DR_REG_ZERO)));
+    PRE(ilist, next_instr, end);
+    /* End of the LR/SC sequence. */
+
+    PRE(ilist, next_instr, XINST_CREATE_jump(dcontext, opnd_create_instr(final)));
+
+    /* Write a non-zero value to dst on fail. */
+    PRE(ilist, next_instr, fail);
+    PRE(ilist, next_instr,
+        XINST_CREATE_load_int(dcontext, dst,
+                              opnd_add_flags(opnd_create_immed_int(1, OPSZ_12b),
+                                             DR_OPND_IMM_PRINT_DECIMAL)));
+
+    PRE(ilist, next_instr, final);
+    PRE(ilist, next_instr,
+        XINST_CREATE_load_int(dcontext, opnd_create_reg(scratch_reg1),
+                              opnd_add_flags(opnd_create_immed_int(-1, OPSZ_12b),
+                                             DR_OPND_IMM_PRINT_DECIMAL)));
+
+    /* Invalidate reservation regardless of success or failure. */
```

Comment:
This belongs before the load_int too?  Writing -1 is what invalidates it?  Why -1?  I would expect 0.  Does it fail if not aligned or sthg?  Please explain in a comment.

Reply:
> This belongs before the load_int too?

Yes, moved it above the load_int.

> Writing -1 is what invalidates it? Why -1? I would expect 0. Does it fail if not aligned or sthg? Please explain in a comment.

Write -1 to the lrsc address slot since -1 is never a valid address, 0 should work too yes. LR/SC does require the address to be aligned, but that will be caught by the CAS sequence. Updated the comment.